sympy/calculus/util.py

```diff
@@ -1631,3 +1632,178 @@ def union(self, other):
+            symbol = list(f.free_symbols)[0]
+
+    if symbol not in f.free_symbols:
+        raise ValueError(filldedent('''
+            argmax of %s cannot be calculated with respect to %s'''%
+            (f, symbol)))
+
+    return _argMaxMin(f, symbol , domain, max = True)
+
+def argmin_real(f, symbol = None, domain = S.Reals):
+    """
+    Returns the set of arguments of minima, i.e., values of `symbol`
+    for which the `f` attains the minimum value in the given domain.
+    Note: This function uses derivatives to calculate the points where the
+    minimum value is achieved, and thus, would not include points where the
+    function is not differentiable. Also, the domain should be a subset of
+    S.Reals
+
+    Parameters
+    ==========
+
+    f: expr
+        The concerned expression
+    symbol: Symbol
+        Variable for which the arguments to be determined.
+    domain: Set
+        The domain over which the arguments are to be determined
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, sin, Interval, S, exp, pi
+    >>> from sympy.calculus.util import argmin_real
+
+    >>> x = Symbol('x')
+    >>> argmin_real(sin(x), x, Interval(-2*pi, pi))
+    FiniteSet(-pi/2)
+
+    >>> argmin_real((x - 1)**2, x, S.Reals)
+    FiniteSet(1)
+
+    >>> argmin_real(exp(x), x, S.Reals)
+    EmptySet
+    """
+    f = _sympify(f)
+    if symbol is None:
+        if len(f.free_symbols) > 1:
+            raise ValueError(filldedent('''
+                Variable for which argmin is to be determined needs to be
+                specified.'''))
+        else:
+            symbol = list(f.free_symbols)[0]
+    if symbol not in f.free_symbols:
+        raise ValueError(filldedent('''
+            argmax of %s cannot be calculated with respect to %s'''%
+            (f, symbol)))
+
+    return _argMaxMin(f, symbol , domain, max = False)
+
+def return_piecewise_solutions(f, x):
+    from sympy.logic.boolalg import And, Or
+
+    solutions = []
+    if type(f) == Piecewise:
+      for arg in f.args:
+          values = []
+          if isinstance(arg[1], (And, Or)):
+              for a in arg[1].args:
+                  a = a.simplify()
+                  values += a.args
+          else:
+              values = arg[1].simplify().args
+          for v in values:
+              if v.is_Number:
+                  solutions.append(v)
+    elif type(f) == Add or type(f) == Mul:
+      for arg in f.args:
+          solutions += return_piecewise_solutions(arg, x)
+
+    return FiniteSet(*solutions)
+
+def _argMaxMin(f, symbol, domain, max):
+    """
+    Helper function for calculuating argmin/argmax
+    """
+    from sympy.solvers.solveset import solveset
+
+    solns = S.EmptySet
+
+    first_deriv = f.diff(symbol)
+    if first_deriv.is_positive:
+        soln1 = FiniteSet(domain.sup)
+        soln2 = FiniteSet(domain.inf)
+        if max:
+            return FiniteSet(list(soln1)[0]) if soln1.is_subset(domain) else S.EmptySet
```

Comment:
How about:
```python
if max:
    point = domain.sup
else:
    point = domain.inf
inrange = domain.contains(point)
if inrange is S.true:
    return FiniteSet(point)
elif inrange is S.false:
    return S.EmptySet
else:
    # This is the None case. What do we do here?
```

Reply:
I don't understand when will contains give ```None``` output?

Reply:
The `contains` method does not give None but in the indeterminate case it won't evaluate to True/False e.g.:
```julia
In [4]: FiniteSet(x, y).contains(z)                                                                                                                           
Out[4]: z ∈ {x, y}

In [5]: type(_)                                                                                                                                               
Out[5]: Contains
```

Reply:
Then we can raise a value error or a not implemented error I guess but i am not certain.

Reply:
If the set is not there we return an empty set. If it is we return the finite set. If we aren't sure we can return an intersection. In fact we can just return an intersection and let that evaluate or not:
```julia
In [14]: FiniteSet(0) & Interval(1, 2)                                                                                                                        
Out[14]: ∅

In [15]: FiniteSet(1) & Interval(1, 2)                                                                                                                        
Out[15]: {1}

In [16]: FiniteSet(x) & Interval(1, 2)                                                                                                                        
Out[16]: [1, 2] ∩ {x}

In [17]: _.subs(x, 0)                                                                                                                                         
Out[17]: ∅
```

Reply:
@oscarbenjamin sure thing will do this.